public/app/plugins/datasource/prometheus/querycache/QueryCache.ts

```diff
@@ -0,0 +1,260 @@
+import {
+  ArrayVector,
+  DataFrame,
+  DataQueryRequest,
+  dateTime,
+  durationToMilliseconds,
+  Field,
+  isValidDuration,
+  parseDuration,
+} from '@grafana/data/src';
+import { amendTable, Table, trimTable } from 'app/features/live/data/amendTimeSeries';
+
+import { PromQuery } from '../types';
+
+// dashboardUID + panelId + refId
+// (must be stable across query changes, time range changes / interval changes / panel resizes / template variable changes)
+type TargetIdent = string;
+
+// query + template variables + interval + raw time range
+// used for full target cache busting -> full range re-query
+type TargetSig = string;
+
+type TimestampMs = number;
+
+type StringInterpolator = (expr: string) => string;
+
+// string matching requirements defined in durationutil.ts
+export const defaultPrometheusQueryOverlapWindow = '10m';
+
+interface TargetCache {
+  sig: TargetSig;
+  prevTo: TimestampMs;
+  frames: DataFrame[];
+}
+
+export interface CacheRequestInfo {
+  requests: Array<DataQueryRequest<PromQuery>>;
+  targSigs: Map<TargetIdent, TargetSig>;
+  shouldCache: boolean;
+}
+
+/**
+ * Get field identity
+ * This is the string used to uniquely identify a field within a "target"
+ * @param field
+ */
+export const getFieldIdent = (field: Field) => `${field.type}|${field.name}|${JSON.stringify(field.labels ?? '')}`;
+
+/**
+ * Get target signature
+ * @param targExpr
+ * @param request
+ * @param targ
+ */
+export function getTargSig(targExpr: string, request: DataQueryRequest<PromQuery>, targ: PromQuery) {
+  return `${targExpr}|${targ.interval ?? request.interval}|${JSON.stringify(request.rangeRaw ?? '')}|${targ.exemplar}`;
+}
+
+/**
+ * NOMENCLATURE
+ * Target: The request target (DataQueryRequest), i.e. a specific query reference within a panel
+ * Ident: Identity: the string that is not expected to change
+ * Sig: Signature: the string that is expected to change, upon which we wipe the cache fields
+ */
+type Milliseconds = number;
+export class QueryCache {
+  private overlapWindow: Milliseconds;
+  constructor(overlapString?: string) {
+    const unverifiedOverlap = overlapString ?? defaultPrometheusQueryOverlapWindow;
+    if (isValidDuration(unverifiedOverlap)) {
+      const duration = parseDuration(unverifiedOverlap);
+      this.overlapWindow = durationToMilliseconds(duration);
+    } else {
+      const duration = parseDuration(defaultPrometheusQueryOverlapWindow);
+      this.overlapWindow = durationToMilliseconds(duration);
+    }
+  }
+
+  cache = new Map<TargetIdent, TargetCache>();
+
+  // can be used to change full range request to partial, split into multiple requests
+  requestInfo(request: DataQueryRequest<PromQuery>, interpolateString: StringInterpolator): CacheRequestInfo {
+    // TODO: align from/to to interval to increase probability of hitting backend cache
```

Comment:
Are we tracking this `todo` somewhere? If not could you please create an issue?

Reply:
i think we are waiting on your assessment if it has any effect ;)

Reply:
I added an item to the epic to address this, I think we can come back to this

Reply:
@leeoniya Ah yes I remember. I forgot that one sorry. As @gtk-grafana suggested let's have it in the epic and come back to it later on